data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/S3InputStream.java

```diff
@@ -0,0 +1,522 @@
+    @Override
+    public int read(ByteBuffer buf) throws IOException {
+        Preconditions.checkState(!closed, "Cannot read: already closed");
+        positionStream();
+
+        int bytesRead = 0;
+        if (buf.hasArray()) {
+            bytesRead = readHeapBuffer(stream, buf);
+        } else {
+            bytesRead = readDirectBuffer(stream, buf, temp);
+        }
+
+        this.bytesCounter.add(bytesRead);
+        return bytesRead;
+    }
+
+    /**
+     * Read {@code buf.remaining()} bytes of data into a {@link ByteBuffer}.
+     * <p>
+     * This method will block until {@code buf.remaining()} bytes are available
+     * to copy into the buffer, or will throw {@link EOFException} if the stream
+     * ends before the buffer is full.
+     *
+     * @param buf a byte buffer to fill with data from the stream
+     * @throws IOException If the underlying stream throws IOException
+     * @throws EOFException If the stream has fewer bytes left than are needed to
+     *                      fill the buffer, {@code buf.remaining()}
+     */
+    @Override
+    public void readFully(ByteBuffer buf) throws IOException {
+        Preconditions.checkState(!closed, "Cannot read: already closed");
+        positionStream();
+
+        int bytesRead = 0;
+        if (buf.hasArray()) {
+            bytesRead = readFullyHeapBuffer(stream, buf);
+        } else {
+            bytesRead = readFullyDirectBuffer(stream, buf, temp);
+        }
+
+        this.bytesCounter.add(bytesRead);
+    }
+
+    /**
+     * Position the stream for reading bytes starting at next offset
+     * @throws IOException if stream cannot be set correctly
+     */
+    private void positionStream() throws IOException {
+        if ((stream != null) && (next == pos)) {
+            // already at specified position
+            return;
+        }
+
+        if ((stream != null) && (next > pos)) {
+            // seeking forwards
+            long skip = next - pos;
+            if (skip <= Math.max(stream.available(), SKIP_SIZE)) {
+                // already buffered or seek is small enough
+                LOG.debug("Read-through seek for {} to offset {}", s3ObjectReference, next);
+                try {
+                    ByteStreams.skipFully(stream, skip);
+                    pos = next;
+                    return;
+                } catch (IOException ignored) {
+                    // will retry by re-opening the stream
+                }
+            }
+        }
+
+        // close the stream and open at desired position
+        LOG.debug("Seek with new stream for {} to offset {}", s3ObjectReference, next);
+        pos = next;
+        openStream();
+    }
+
+    /**
+     * Open the stream to the S3 object
+     * @throws IOException if the stream cannot be opened.
+     */
+    private void openStream() throws IOException {
+        final GetObjectRequest request = this.getObjectRequestBuilder
+                .range(String.format("bytes=%s-", pos))
```

Comment:
Are we always needing to get byte ranges? Can we just get a single stream for the use-cases where this isn't needed?

Reply:
There is no penalty of specifying `byte=0-`. That's simpler than checking if pos=0 and not passing in the range if it is == 0.